packages/core/mesh/edge-client/src/edge-client.ts

```diff
@@ -98,13 +94,32 @@ export class EdgeClient extends Resource implements EdgeConnection {
     if (identity.identityKey !== this._identity.identityKey || identity.peerKey !== this._identity.peerKey) {
       log('Edge identity changed', { identity, oldIdentity: this._identity });
       this._identity = identity;
+      this._closeCurrentConnection(new EdgeIdentityChangedError());
       this._persistentLifecycle.scheduleRestart();
     }
   }
 
-  public addListener(listener: MessageListener): () => void {
-    this._listeners.add(listener);
-    return () => this._listeners.delete(listener);
+  public onMessage(listener: MessageListener): () => void {
+    this._messageListeners.add(listener);
+    return () => this._messageListeners.delete(listener);
+  }
+
+  public onReconnected(listener: () => void): () => void {
```

Comment:
I would call this `onConnected` or `withConnection` and add a brief jsdoc to explain the behaviour. The name `onReconnected` suggests that the callback will not be called if it is already connected